### DISCUSSION TITLE
Drawing of paths with separate segments in a single layer?

### DISCUSSION THREAD

**Discussion by moving-bits** [mapsforge/vtm#970] [Q&A] [answered]
_(Continuing discussion from https://github.com/mapsforge/vtm/discussions/969#discussioncomment-4011897 in a separate thread for this topic)_

Original question:

I have a couple of paths to draw (eg.: each path represents an accumulated track history for a specific month).

- Each path can consist of several unrelated segments (eg.: track history for a single day),
- each segment consisting of several points (eg.: the places you have visited on that specific day) - this should be an actual path with connected elements.

I'm using a `PathLayer` to display this, but I have not been able to show the segments unconnected, so I'm currently putting each segment into its own layer. Surely there must be a better way to do this.

Reply given in https://github.com/mapsforge/vtm/discussions/969#discussioncomment-4012065:

> GPX tracks with multiple track segments: need a path layer per track segment (Mapsforge + VTM).
> Map libraries do not care if 2 lines are "related", if not connected then they are separate line layers.

For Mapsforge we currently use something like this:
```
c.path = AndroidGraphicFactory.INSTANCE.createPath();
ArrayList<Geopoint> segment = segmentIterator.next();
while (segment != null) {
    final Iterator<Geopoint> geopointIterator = segment.iterator();
    Geopoint geopoint = geopointIterator.next();
    c.path.moveTo((float) (MercatorProjection.longitudeToPixelX(geopoint.getLongitude(), mapSize) - topLeftPoint.x), (float) (MercatorProjection.latitudeToPixelY(geopoint.getLatitude(), mapSize) - topLeftPoint.y));

    while (geopointIterator.hasNext()) {
        geopoint = geopointIterator.next();
        c.path.lineTo((float) (MercatorProjection.longitudeToPixelX(geopoint.getLongitude(), mapSize) - topLeftPoint.x), (float) (MercatorProjection.latitudeToPixelY(geopoint.getLatitude(), mapSize) - topLeftPoint.y));
    }
    segment = segmentIterator.hasNext() ? segmentIterator.next() : null;
}
```
which separates segments nicely, thus having to use only a single layer for the whole track (or even multiple tracks).

I'm looking for something similar for VTM to be able to stick with a fixed number of layers, as
- layer index defines visibility and track layer is not top-most for me
- number of segments is unknown to me upfront, and can vary during lifespan
- layer index values can have no gaps.

Any other type of layer I can choose by which I can achieve something similar like in our Mapsforge code?

**Comment by devemux86** [comment]
Can study the `vtm-jts` module and create any extra custom layers.

In any case map engines are different, so must not expect same API.
Currently the general rule is that separate lines use separate layers.

**Comment by moving-bits** [comment, marked as answer]
Found a simple solution meanwhile:

Instead of
```
final PathLayer trackLayer = new PathLayer(map, color, width);
map.layers().add(trackLayer);
trackLayer.setPoints(segment);
```

I'm using this now:
```
final Style pathStyle = Style.builder().strokeWidth(width).strokeColor(color).build();
final LineDrawable path = new LineDrawable(segment, pathStyle);
trackLayer.add(path);
```
(with `trackLayer` being the `VectorLayer` holding all my tracks now).

Thanks for your help pointing me into the right direction!

**Reply by devemux86** [reply]
Thanks for writing the solution!
It's nice that there is already an alternative.
